doc/developers-guide/logging.md

```diff
@@ -0,0 +1,103 @@
+# Logging
+
+To use logger in your module, include
+
+```erlang
+-include("mongoose_logger.hrl").
+```
+
+or
+
+```erlang
+-include("mongoose.hrl").
+```
+
+# Logging macroses
+
+There are several macroses:
+
+```erlang
+DEBUG("event=debug_event info=~1000p", [Arg]),
+INFO_MSG("event=info_event info=~1000p", [Arg]),
+ERROR_MSG("event=error_event reason=~1000p", [Arg]),
+WARNING_MSG("event=warning_event reason=~1000p", [Arg]),
+```
+
+Use them to log on corresponding log level.
+
+Don't use `ERROR_MSG` for cases that are not errors.
+
+# Logging levels
+
+A system operator can choose log level by setting loglevel in `mongooseim.cfg`.
+
+- level 5 - debug
+- level 4 - info
+- level 3 - warning
+- level 2 - error
+- level 1 - critical
+- level 0 - none
+
+While there is a macro to log on level 1, we rarely use it.
+
+If an user sets log level 5, than he would see all messages in logs.
+
+Levels 3 or 2 are usually used for production systems.
+
+
+# Logging format
+
+We use modified [logfmt](https://brandur.org/logfmt) format.
+
+This format is Splunk and ELK friendly.
+
+`event=something_interesting` field is required.
+
+`reason=~1000p` field is commonly used.
+
+```erlang
+    ?ERROR_MSG("event=check_password_failed "
+               "reason=~p user=~ts", [Error, LUser]),
+
+    try ...
+    catch
+        Class:Reason ->
+            Stacktrace = erlang:get_stacktrace(),
+            ?ERROR_MSG("event=check_password_failed "
+                       "reason=~p:~p user=~ts stacktrace=~1000p",
+                       [Class, Reason, LUser, Stacktrace]),
+            erlang:raise(Class, Reason, Stacktrace)
+    end
+```
+
+`user=~ts` is often used too.
+
+
+A common way to name event is `event=function_name_failed`.
```

Comment:
Let's not advise it. Naming things is hard and our functions rarely have good names :P We can check which function has failed by enabling lager's `function` metadata.

Reply:
> A common way to name event is `event=function_name_failed`, but use common sense :)